google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/inventory.rb

```diff
@@ -0,0 +1,106 @@
+# Copyright 2017 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "monitor"
+
+module Google
+  module Cloud
+    module Pubsub
+      class Subscriber
+        ##
+        # @private
+        class Inventory
+          include MonitorMixin
+
+          attr_reader :stream, :limit
+
+          def initialize stream, limit
+            @stream = stream
+            @limit = limit
+            @_ack_ids = []
+            @wait_cond = new_cond
+
+            super()
+          end
+
+          def ack_ids
+            @_ack_ids
+          end
+
+          def add *ack_ids
+            ack_ids = Array(ack_ids).flatten
+            synchronize do
+              @_ack_ids += ack_ids
+              unless @stopped
+                @background_thread ||= Thread.new { background_run }
+              end
+            end
+          end
+
+          def remove *ack_ids
+            ack_ids = Array(ack_ids).flatten
+            synchronize do
+              @_ack_ids -= ack_ids
+              if @_ack_ids.empty?
+                if @background_thread
+                  @background_thread.kill
```

Comment:
You're using both `@stopped` and `Thread#kill`. Best to stick with one or the other. And I think it's better to choose `@stopped` (i.e. avoid `Thread#kill`) unless you actually need the semantics of "immediately and brutally interrupt any currently running `stream.delay_inventory!` call if the last ack_ids are removed."

Reply:
Yeah, I was thinking the same thing as I was moving the code to the new file. I don't love the approach of killing and restarting threads over and over.

I think this is a situation where both approaches were used in order to get it working and once it did we just stopped touching it. I'll poke at it a bit and see what I can do to keep one thread active and kill/create.

Reply:
Updated to no longer kill a thread. Condition variables are now used to pause background activity.